### PR TITLE
Fix infinite loop stuck

### PR DESCRIPTION
**Pull request by Arpita-Jaiswal**
This PR introduces the `ftd::interpreter::InterpreterState::detect_cycle` function to detect cycles in the component processing sequence and raises an error if a cycle is found. It prevents infinite loops caused by cyclic dependencies during component processing. Components may be processed multiple times if they have unprocessed dependencies or if it's stuck in an infinite loop. However, this function ensures that no component is stuck in a recursive cycle by distinguishing between legitimate reprocessing (due to unresolved dependencies) and infinite loops (due to cyclic dependencies).

## Theory of cycle detection

A cycle is detected when a repeating pattern of components appears in the `in-process` list, for example:

```
 a -> b -> c -> d -> e -> c -> d -> e -> c -> d -> e
           ^              ^              ^
```

In this case, `c -> d -> e` is the repeating pattern, indicating a cyclic dependency that will never resolve.

##  How it works:
1. **Remove Already Processed Components**: The function begins by clearing any components that have already been fully processed from the `in-process` list to avoid redundant checks.

2. **Skip Already Processed Nodes**: It skips AST nodes that are either invocations, imports, or have already been processed (tracked via `self.bag`). This prevents unnecessary re-processing.

3. **Handling Multiple In-Process Entries**: If a component appears multiple times to be processed, it could either be due to unprocessed dependencies or a cyclic dependency (infinite loop). To handle this, we use `number_of_scan`. Every time a component is re-processed because of unprocessed dependencies, the `number_of_scan` is incremented. This allows us to differentiate between normal re-processing and an infinite loop.

4. **Cycle Detection**: The function then checks for cycles by comparing the current `in-process` sequence with its prior state:
    - It compares the last half of the `in-process` list to the preceding half to check for repeating 
      patterns (which would indicate a cycle).
    - If a cycle is detected, an error is raised by calling `self.construct_cycle_error(i)`.

5. **Completion**: If no cycle is detected, the function proceeds normally, and the component is processed without error.

### CODE REVIEW

**Comment by amitu** [comment]
@Arpita-Jaiswal tests are failing

**Comment by Arpita-Jaiswal** [comment]
> @Arpita-Jaiswal tests are failing

@amitu The PR is not complete yet.

**Comment by amitu** [comment]
@Arpita-Jaiswal can you also take a look at this one: https://github.com/fastn-stack/fastn/issues/552

Also I think there was another issue related to partial renaming of component, eg component has been renamed, but old component name is still referenced in the body of component, causing infinite loop.

File: ftd/t/interpreter/32-recursion.ftd [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+-- ftd.column:
+
+-- ftd.text: $print-tree.t.name
+
+-- print-tree: $obj
+for: $obj in $print-tree.t.children
+
+-- end: ftd.column
+
+-- end: print-tree
```
Comment (amitu) [inline ftd/t/interpreter/32-recursion.ftd:35]: All these red signs, get rid of them. Add new line at the end of each file that doesn't have them.

**Review by amitu — APPROVED** [review]

**Comment by amitu** [comment]
@Arpita-Jaiswal do a version bump, merge this PR and create a release.

**Pull request merged**